solutions/usaco-838.mdx

```diff
@@ -2,23 +2,105 @@
 id: usaco-838
 source: USACO Gold 2018 Open
 title: Milking Order
-author: Qi Wang, Melody Yu
+author: Qi Wang, Melody Yu, Neo Wang
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_milkorder_gold_open18.html)
 
 ## Video Solution
 
 Note: The video solution might not be the same as other solutions. Code in C++.
+
 <Youtube id="OaL9vEkShyk" />
 
+## Implementation
+
 **Time Complexity:** $\mathcal{O}((E + N \log N) \log M)$
 
+We can binary search on the last valid value of $M$ such that the ordering can
+be satisfied. For each iteration of the binary search, we only check if the
+interval of times $[0, x)$ is valid. Finally, to
+find the lexigraphically smallest TopoSort, we use a `priority_queue` to process
+the ids from least to greatest.
+
 <LanguageSection>
 
-<JavaSection>
+<CPPSection>
+
+```cpp
+//CodeSnip{CPP Short Template}
+//BeginCodeSnip{Topographical Sort}
+// /**
+//  * Description: sorts vertices such that if there exists an edge x->y, then x goes before y
+//  * Source: KACTL
+//  * Verification: https://open.kattis.com/problems/quantumsuperposition
+//  */
+
+struct TopoSort {
+	int N; vi in, res;
+	vector<vi> adj;
+	void init(int _N) { N = _N; in.resize(N); adj.resize(N); }
+	void ae(int x, int y) { adj[x].pb(y), in[y] ++; }
+	bool sort() {
+		priority_queue<int, vi, greater<int>> todo;
+		for(int i = 1; i < N; i++) if (!in[i]) todo.push(i);
+		while (sz(todo)) {
+			int x = todo.top(); todo.pop(); res.pb(x);
+			for(const auto &i : adj[x]) if (!(--in[i])) todo.push(i);
+		}
+		return sz(res) == N - 1;
+	}
+};
+//EndCodeSnip{Topographical Sort}
+
+const int mx = 2e5+1;
+
+vi ret;
+vector<vi> order;
+int n, m;
+
+bool check(int x) {
+	TopoSort T; T.init(n + 1);
+	for(int i = 0; i < x; i++) {
```

Comment:
put a space after the keyword please
like
```suggestion
	for (int i = 0; i < x; i++) {
```
you've done it elsewhere so just be consistent